#📌 Extracción

#### Diccionario de datos

- `customerID`: número de identificación único de cada cliente
- `Churn`: si el cliente dejó o no la empresa
- `gender`: género (masculino y femenino)
- `SeniorCitizen`: información sobre si un cliente tiene o no una edad igual o mayor a 65 años
- `Partner`: si el cliente tiene o no una pareja
- `Dependents`: si el cliente tiene o no dependientes
- `tenure`: meses de contrato del cliente
- `PhoneService`: suscripción al servicio telefónico
- `MultipleLines`: suscripción a más de una línea telefónica
- `InternetService`: suscripción a un proveedor de internet
- `OnlineSecurity`: suscripción adicional de seguridad en línea
- `OnlineBackup`: suscripción adicional de respaldo en línea
- `DeviceProtection`: suscripción adicional de protección del dispositivo
- `TechSupport`: suscripción adicional de soporte técnico, menor tiempo de espera
- `StreamingTV`: suscripción de televisión por cable
- `StreamingMovies`: suscripción de streaming de películas
- `Contract`: tipo de contrato
- `PaperlessBilling`: si el cliente prefiere recibir la factura en línea
- `PaymentMethod`: forma de pago
- `Charges.Monthly`: total de todos los servicios del cliente por mes
- `Charges.Total`: total gastado por el cliente

#👀 Visualizacion de los datos

In [102]:
import pandas as pd

url = "https://raw.githubusercontent.com/ingridcristh/challenge2-data-science-LATAM/main/TelecomX_Data.json"

df = pd.read_json(url)


df.head()


customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [103]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [104]:
datos = pd.unique(df['customerID'])
datos1 = pd.unique(df['Churn'])

print(datos)
print(datos1)


['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
['No' 'Yes' '']


#🔧 Transformación

## Normalizacionde de datos anidados

In [105]:
datos_customer = pd.json_normalize(df['customer'])
datos_phone = pd.json_normalize(df['phone'])
datos_internet = pd.json_normalize(df['internet'])
datos_account = pd.json_normalize(df['account'])

df_limpio = pd.concat([df[['customerID', 'Churn']], datos_customer, datos_phone, datos_internet, datos_account], axis=1)

df_limpio


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

##Estandarizacion de los datos

In [106]:
df_limpio.isnull().sum()


customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

In [120]:
df_limpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7043 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   float64
 4   Partner           7267 non-null   Int64  
 5   Dependents        7267 non-null   Int64  
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   Int64  
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 




 3   SeniorCitizen     7267 non-null   int64 bool

 20  Charges.Total     7267 non-null   object Float


In [108]:
df_limpio['Charges.Total'] = pd.to_numeric(df_limpio['Charges.Total'], errors='coerce')

In [109]:
df_limpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [110]:
for col in df_limpio.columns:
    print(f'{col}: {pd.unique(df_limpio[col])}')

customerID: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
Churn: ['No' 'Yes' '']
gender: ['Female' 'Male']
SeniorCitizen: [0 1]
Partner: ['Yes' 'No']
Dependents: ['Yes' 'No']
tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 11 37 49 66
 67 20 43 59 12 27  2 25 29 14 35 64 39 40  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 51 31 36 17 18 38 42
  0]
PhoneService: ['Yes' 'No']
MultipleLines: ['No' 'Yes' 'No phone service']
InternetService: ['DSL' 'Fiber optic' 'No']
OnlineSecurity: ['No' 'Yes' 'No internet service']
OnlineBackup: ['Yes' 'No' 'No internet service']
DeviceProtection: ['No' 'Yes' 'No internet service']
TechSupport: ['Yes' 'No' 'No internet service']
StreamingTV: ['Yes' 'No' 'No internet service']
StreamingMovies: ['No' 'Yes' 'No internet service']
Contract: ['One year' 'Month-to-month' 'Two year']
PaperlessBilling: ['Yes' 'No']
PaymentMethod: ['Mailed check' 'Electronic check' 'Cr

In [111]:
import numpy as np

df_limpio['SeniorCitizen'] =  df_limpio['SeniorCitizen'].astype(np.float64)

df_limpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   float64
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [112]:
df_limpio.applymap(lambda x: str(x).strip() in [""]).sum()


/tmp/ipython-input-1262798319.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_limpio.applymap(lambda x: str(x).strip() in [""]).sum()


customerID            0
Churn               224
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod         0
Charges.Monthly       0
Charges.Total         0
dtype: int64

In [113]:
df_limpio.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female            0.0     Yes        Yes       9   
1  0003-MKNFE    No    Male            0.0      No         No       9   
2  0004-TLHLJ   Yes    Male            0.0      No         No       4   
3  0011-IGKFF   Yes    Male            1.0     Yes         No      13   
4  0013-EXCHZ   Yes  Female            1.0     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6         593.30  
1               No      Mailed check            59.9         542.40  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9         267.40  

[5 rows x 21 columns]

In [114]:
df_limpio['Churn'] = df_limpio['Churn'].replace('',pd.NA)
print(df_limpio['Churn'].unique())
print(df_limpio['Churn'].isnull().sum())

['No' 'Yes' <NA>]
224


In [115]:
df_limpio


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female            0.0     Yes        Yes       9   
1     0003-MKNFE    No    Male            0.0      No         No       9   
2     0004-TLHLJ   Yes    Male            0.0      No         No       4   
3     0011-IGKFF   Yes    Male            1.0     Yes         No      13   
4     0013-EXCHZ   Yes  Female            1.0     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female            0.0      No         No      13   
7263  9992-RRAMN   Yes    Male            0.0     Yes         No      22   
7264  9992-UJOEL    No    Male            0.0      No         No       2   
7265  9993-LHIEB    No    Male            0.0     Yes        Yes      67   
7266  9995-HOTOH    No    Male            0.0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60         593.30  
1                  No      Mailed check           59.90         542.40  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90         267.40  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15         742.90  
7263              Yes  Electronic check           85.10        1873.70  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00        3707.60  

[7267 rows x 21 columns]

In [116]:
# Crear columna ChargesDaily a partir de Charges.Monthly
df_limpio['ChargesDaily'] = df_limpio['Charges.Monthly'] / 30

# Verificar los primeros valores
df_limpio['ChargesDaily'] = (df_limpio['Charges.Monthly'] / 30).round(2)
print(df_limpio[['Charges.Monthly', 'ChargesDaily']].head())


   Charges.Monthly  ChargesDaily
0             65.6          2.19
1             59.9          2.00
2             73.9          2.46
3             98.0          3.27
4             83.9          2.80


In [117]:
# Lista de columnas convertidas (opcional)
columnas_convertidas = []

# Recorremos todas las columnas tipo object, excepto 'Churn'
for col in df_limpio.select_dtypes(include='object').columns:
    if col == 'Churn':
        continue  # Saltar esta columna

    valores_unicos = df_limpio[col].dropna().unique()

    if set(valores_unicos) == {'Yes', 'No'}:
        df_limpio[col] = df_limpio[col].map({'Yes': 1, 'No': 0}).astype('Int64')
        columnas_convertidas.append(col)

# Mostrar columnas que fueron transformadas
print("Columnas convertidas a 0/1:", columnas_convertidas)


Columnas convertidas a 0/1: ['Partner', 'Dependents', 'PhoneService', 'PaperlessBilling']


In [118]:
df_limpio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7043 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   float64
 4   Partner           7267 non-null   Int64  
 5   Dependents        7267 non-null   Int64  
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   Int64  
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


#üìä Carga y an√°lisis

##Analisis descriptivo

In [ ]:
df_limpio.describe()

## Estudio de Evasion

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Gráfico de barras incluyendo <NA> (con corrección de warning)
plt.figure(figsize=(6, 4))
sns.countplot(data=df_limpio, x='Churn', hue='Churn', palette='Set2', legend=False)
plt.title('Distribución de Clientes según Churn (incluye datos faltantes)')
plt.xlabel('¬øCliente dado de baja?')
plt.ylabel('Cantidad de clientes')
plt.show()

# Gr√°fico de pastel incluyendo <NA>
churn_counts = df_limpio['Churn'].value_counts(dropna=False)
labels = ['No', 'Yes', 'Desconocido']
plt.figure(figsize=(5, 5))
plt.pie(churn_counts, labels=labels, autopct='%1.1f%%', startangle=90,
        colors=['#66c2a5', '#fc8d62', '#999999'])
plt.title('Proporción de Clientes según Churn (incluye datos faltantes)')
plt.axis('equal')
plt.show()



In [ ]:
# Crear tabla cruzada
tabla_genero = df_limpio[df_limpio['Churn'].notna()].groupby(['Churn', 'gender']).size().unstack()

# Gr√°fico de barras apiladas
tabla_genero.plot(kind='bar', stacked=True, figsize=(6, 4), colormap='Set2')
plt.title('Composición de género por categoría de Churn')
plt.xlabel('¬øCliente dado de baja?')
plt.ylabel('Cantidad de clientes')
plt.legend(title='Género')
plt.tight_layout()
plt.show()


# Normalizar para obtener proporciones (porcentaje por grupo de Churn)
tabla_genero_pct = tabla_genero.div(tabla_genero.sum(axis=1), axis=0) * 100

# Crear gr√°fico con etiquetas
fig, ax = plt.subplots(figsize=(6, 4))
colors = plt.get_cmap('Set2').colors
tabla_genero_pct.plot(kind='bar', stacked=True, ax=ax, color=colors)

# Agregar etiquetas de porcentaje dentro de las barras
for i, (idx, row) in enumerate(tabla_genero_pct.iterrows()):
    y_offset = 0
    for j, (col, val) in enumerate(row.items()):
        if val > 3:  # Oculta etiquetas muy pequeñas para evitar ruido visual
            ax.text(i, y_offset + val / 2, f"{val:.1f}%", ha='center', va='center', fontsize=9)
        y_offset += val

# Estética del gráfico
ax.set_title('Composición porcentual de género por categoría de Churn')
ax.set_ylabel('Porcentaje (%)')
ax.set_xlabel('¬øCliente dado de baja?')
ax.legend(title='Género', loc='upper right')
plt.tight_layout()
plt.show()


In [ ]:
# Tabla cruzada normalizada
tabla_contrato = df_limpio[df_limpio['Churn'].notna()].groupby(['Churn', 'Contract']).size().unstack()
tabla_contrato_pct = tabla_contrato.div(tabla_contrato.sum(axis=1), axis=0) * 100

# Crear gr√°fico
fig, ax = plt.subplots(figsize=(7, 4))
colors = plt.get_cmap('Set3').colors
bars = tabla_contrato_pct.plot(kind='bar', stacked=True, ax=ax, color=colors)

# Añadir etiquetas de porcentaje sobre las barras
for i, (idx, row) in enumerate(tabla_contrato_pct.iterrows()):
    y_offset = 0
    for j, (col, val) in enumerate(row.items()):
        if val > 2:  # Evita superposición con valores muy pequeños
            ax.text(i, y_offset + val / 2, f"{val:.1f}%", ha='center', va='center', fontsize=9)
        y_offset += val

# Estética
ax.set_title('Composición porcentual de tipo de contrato por categoría de Churn')
ax.set_xlabel('¬øCliente dado de baja?')
ax.set_ylabel('Porcentaje (%)')
ax.legend(title='Tipo de contrato', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.tight_layout()
plt.show()

In [ ]:
# Tabla cruzada normalizada
tabla_pago = df_limpio[df_limpio['Churn'].notna()].groupby(['Churn', 'PaymentMethod']).size().unstack()
tabla_pago_pct = tabla_pago.div(tabla_pago.sum(axis=1), axis=0) * 100

# Crear gr√°fico
fig, ax = plt.subplots(figsize=(8, 4))
colors = plt.get_cmap('Pastel1').colors
tabla_pago_pct.plot(kind='bar', stacked=True, ax=ax, color=colors)

# Añadir etiquetas de porcentaje
for i, (idx, row) in enumerate(tabla_pago_pct.iterrows()):
    y_offset = 0
    for j, (col, val) in enumerate(row.items()):
        if val > 3:
            ax.text(i, y_offset + val / 2, f"{val:.1f}%", ha='center', va='center', fontsize=8)
        y_offset += val

# Estética
ax.set_title('Composición porcentual de método de pago por categoría de Churn')
ax.set_xlabel('¬øCliente dado de baja?')
ax.set_ylabel('Porcentaje (%)')
ax.legend(title='Método de pago', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.tight_layout()
plt.show()


In [ ]:
# Asegurarse de que los datos sean numéricos (en caso de que 'Charges.Total' sea string)
df_limpio['Charges.Total'] = pd.to_numeric(df_limpio['Charges.Total'], errors='coerce')

# Filtrar solo filas sin valores faltantes en Churn
df_filtrado = df_limpio[df_limpio['Churn'].notna()]

# Listado de variables numéricas a analizar
variables_numericas = ['Charges.Total', 'tenure']

for var in variables_numericas:
    plt.figure(figsize=(6, 4))
    sns.boxplot(data=df_filtrado, x='Churn', y=var, palette='Set2')
    plt.title(f'Distribución de {var} según Churn')
    plt.xlabel('¬øCliente dado de baja?')
    plt.ylabel(var)
    plt.tight_layout()
    plt.show()


#📄Informe final

# üìÖ Informe Final: An√°lisis de Evasi‚èön de Clientes - TelecomX LATAM

---

## 🔍 Introducción

El presente informe aborda el análisis de datos de clientes de **TelecomX LATAM**, con el objetivo de estudiar el fenómeno de *Churn* (evasión o baja de clientes). Comprender las características de los clientes que abandonan la compañía es fundamental para diseñar estrategias de retención más efectivas.

---

## ⚙️ Limpieza y Tratamiento de Datos

Se trabajó a partir de un archivo JSON alojado en GitHub. El dataset estaba estructurado con datos anidados que fueron normalizados en distintas tablas: `customer`, `phone`, `internet` y `account`. Se fusionaron en un DataFrame final llamado `df_limpio`.

Pasos realizados:

- Conversión de columnas categóricas binarias ('Yes'/'No') a valores `0` y `1`.
- Reemplazo de cadenas vacías en `Churn` por `NaN`.
- Conversión de `Charges.Total` a tipo numérico.
- Creación de una nueva columna `ChargesDaily` = `Charges.Monthly / 30`.

---

## üìä An√°lisis Exploratorio de Datos

### 📊 Distribución General de Churn

| Churn | Cantidad | Porcentaje (%) |
|-------|----------|----------------|
| No    | 5174     | 73.46%         |
| Yes   | 1869     | 26.54%         |

---

### 🧍‍♂️ Composición de Género por Churn

| Churn | Female | Male  |
|-------|--------|-------|
| No    | 49.27% | 50.73%|
| Yes   | 50.24% | 49.76%|

---

### 📄 Tipo de Contrato por Churn

| Churn | Month-to-month | One year | Two year |
|-------|----------------|----------|----------|
| No    | 42.91%         | 25.26%   | 31.83%   |
| Yes   | 88.55%         | 8.88%    | 2.57%    |

---

### 💳 Método de Pago por Churn

| Churn | Bank transfer | Credit card | Electronic check | Mailed check |
|-------|----------------|--------------|-------------------|----------------|
| No    | 24.86%         | 24.93%       | 25.01%            | 25.20%         |
| Yes   | 13.80%         | 12.41%       | 57.30%            | 16.48%         |

---

### 📈 Estadísticas Numéricas por Churn

#### Tenure (tiempo en meses con la compañía)

| Churn | Count | Mean  | Std   | Min | Max |
|-------|-------|-------|-------|-----|-----|
| No    | 5174  | 37.57 | 24.11 | 0   | 72  |
| Yes   | 1869  | 17.98 | 19.53 | 1   | 72  |

#### Charges.Total (Monto total gastado)

| Churn | Count | Mean  | Std   | Min | Max |
|-------|-------|-------|-------|-----|-----|
| No    | 5174  | 0.13  | 0.33  | 0.0 | 1.0 |
| Yes   | 1869  | 0.25  | 0.44  | 0.0 | 1.0 |

---

## üåê Observaciones Visuales (resumen)

- Clientes que se dan de baja tienden a tener **menos tiempo con la empresa**.
- Contratos `Month-to-month` son muy comunes en clientes con evasión.
- El uso de `Electronic check` est√° claramente asociado con mayor churn.

---

## 🤝 Conclusiones e Insights

- La mayoría de los clientes que abandonan tienen **poco tiempo con la compañía**.
- El contrato mensual representa **casi el 90% de los casos de churn**.
- El método de pago `Electronic check` tiene una relación directa con la evasión.

---

## üí° Recomendaciones

1. **Fidelizar clientes nuevos** con ofertas o soporte adicional en sus primeros meses.
2. **Incentivar contratos a largo plazo** con descuentos progresivos.
3. **Desalentar el uso de 'Electronic check'** ofreciendo otros métodos más confiables.
4. Desarrollar un **modelo predictivo de churn** con estas variables clave para anticiparse a futuras bajas.

---

> 📌 Este informe entrega las bases para diseñar estrategias orientadas a reducir la fuga de clientes y mejorar la retención en **TelecomX LATAM**.


##EXTRA:  Análisis de correlación entre variables

In [ ]:
# Crear copia y limpieza
df_corr = df_limpio.copy()

# Convertir columnas de servicios de texto a 0/1
servicios = ['OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies']
df_corr[servicios] = df_corr[servicios].replace({'No internet service': 'No'})
df_corr['MultipleLines'] = df_corr['MultipleLines'].replace({'No phone service': 'No'})

# Convertir a numérico binario
binarias = servicios + ['MultipleLines']
df_corr[binarias] = df_corr[binarias].replace({'Yes': 1, 'No': 0})

# Crear columna total de servicios contratados
df_corr['ServiciosContratados'] = df_corr[binarias].sum(axis=1)

# Crear columna Churn_num
df_corr['Churn_num'] = df_corr['Churn'].map({'No': 0, 'Yes': 1})

# Filtrar datos v√°lidos
df_corr_clean = df_corr[df_corr['Churn_num'].notna()]

# Seleccionar variables relevantes
variables_corr = [
    'tenure', 'Charges.Monthly', 'Charges.Total', 'ChargesDaily',
    'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'PaperlessBilling',
    'ServiciosContratados'
] + binarias

# Calcular correlación con churn y mostrar como DataFrame
cor_churn = df_corr_clean[variables_corr + ['Churn_num']].corr()['Churn_num'].drop('Churn_num').sort_values(ascending=False)

# Mostrar resultado
cor_churn_df = cor_churn.to_frame(name='Correlación con Churn')
display(cor_churn_df)


In [ ]:
# Crear copia y limpieza
df_corr = df_limpio.copy()

# Convertir columnas de servicios de texto a 0/1
servicios = ['OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies']
df_corr[servicios] = df_corr[servicios].replace({'No internet service': 'No'})
df_corr['MultipleLines'] = df_corr['MultipleLines'].replace({'No phone service': 'No'})

# Convertir a numérico binario
binarias = servicios + ['MultipleLines']
df_corr[binarias] = df_corr[binarias].replace({'Yes': 1, 'No': 0})

# Crear columna total de servicios contratados
df_corr['ServiciosContratados'] = df_corr[binarias].sum(axis=1)

# Crear columna Churn_num
df_corr['Churn_num'] = df_corr['Churn'].map({'No': 0, 'Yes': 1})

# Filtrar datos v√°lidos
df_corr_clean = df_corr[df_corr['Churn_num'].notna()]

# Seleccionar variables relevantes
variables_corr = [
    'tenure', 'Charges.Monthly', 'Charges.Total', 'ChargesDaily',
    'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'PaperlessBilling',
    'ServiciosContratados'
] + binarias

# Calcular correlación con churn y mostrar como DataFrame
cor_churn = df_corr_clean[variables_corr + ['Churn_num']].corr()['Churn_num'].drop('Churn_num').sort_values(ascending=False)

# Mostrar resultado
cor_churn_df = cor_churn.to_frame(name='Correlación con Churn')
display(cor_churn_df)


## 📈 Análisis de Correlación con Churn

A continuación se muestra la correlación de diversas variables con respecto a la variable `Churn` (evasión de clientes), expresada como valores entre `-1` (correlación negativa fuerte) y `+1` (correlación positiva fuerte).

| Variable             | Correlación con Churn |
|----------------------|------------------------|
| ChargesDaily         | 0.193                  |
| Charges.Monthly      | 0.193                  |
| PaperlessBilling     | 0.192                  |
| SeniorCitizen        | 0.151                  |
| Charges.Total        | 0.151                  |
| StreamingTV          | 0.063                  |
| StreamingMovies      | 0.061                  |
| MultipleLines        | 0.040                  |
| PhoneService         | 0.012                  |
| DeviceProtection     | -0.066                 |
| ServiciosContratados | -0.070                 |
| OnlineBackup         | -0.082                 |
| Partner              | -0.150                 |
| Dependents           | -0.164                 |
| TechSupport          | -0.165                 |
| OnlineSecurity       | -0.171                 |
| tenure               | -0.352                 |

---

## 🔍 Conclusiones de la correlación

### üìâ 1. `tenure` (tiempo con la empresa): **la variable m√°s inversamente correlacionada con churn** (`-0.35`)
> Cuanto más tiempo lleva un cliente, menos probable es que se dé de baja. Indica lealtad o satisfacción acumulada.

### 🛡️ 2. Variables como `OnlineSecurity`, `TechSupport` y `OnlineBackup` tienen **correlaciones negativas moderadas**:
> Los clientes que usan más servicios tienden a mantenerse. Sumar servicios podría ser una estrategia de retención.

### 💳 3. `PaperlessBilling` y `Charges.Monthly/Daily`: **correlación positiva moderada**
> Los clientes que pagan electrónicamente y que pagan más tienden a darse de baja ligeramente más. Podrían ser más exigentes o sensibles a precios altos.

### 👵 4. `SeniorCitizen`: **correlación positiva leve** (`0.15`)
> Puede reflejar desconexión con tecnología, dificultades económicas o menor fidelización.

### 👨‍👩‍👧 5. `Partner` y `Dependents`: **negativamente correlacionados**
> Tener pareja o dependientes puede reflejar mayor estabilidad o compromiso, disminuyendo la evasión.

---

## ✅ Conclusión general

El churn est√° m√°s fuertemente asociado con:

- Poco tiempo como cliente.
- Falta de servicios de soporte técnico o seguridad.
- Ser cliente con facturación electrónica.
- No tener responsabilidades familiares.

Estas correlaciones no implican causalidad, pero ofrecen **señales valiosas para diseñar estrategias de retención**.

